File: java/145-BinaryTreePostorderTraversal.java
/*

Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?

*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List <Integer> res = new ArrayList<>();
        TreeNode dummy = new TreeNode(0);
        dummy.left = root;
        TreeNode prev = null;
        TreeNode curr = dummy;
        while (curr != null) {
            if(curr.left == null) { // no left, go right
                curr = curr.right;
            } 
            else { 
                prev  = curr.left;
                while(prev.right!=null && prev.right!=curr) prev = prev.right;
                if(prev.right == null) { // create threading and go left
                    prev.right = curr;
                    curr = curr.left;
                } 
                else { //remove threading and go right
                    List<Integer> temp = trace_back(curr.left, prev);
                    for(Integer n: temp) res.add(n);
                    prev.right = null;
                    curr = curr.right;
                }
            }
        }
        return res;
    }
    private List<Integer> trace_back(TreeNode from, TreeNode to) {
        List<Integer> temp = new ArrayList<>();
        TreeNode curr = from;
        while(curr!=to) {
            temp.add(curr.val);
            curr = curr.right;
        }
        temp.add(to.val);
        Collections.reverse(temp);
        return temp;
    }
}

class Solution1 {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if(root == null) return res;
        Stack<Node> s = new Stack<>();
        Node t = new Node(root, false);
        s.push(t);
        while(!s.isEmpty()) {
            Node temp = s.pop();
            if(temp.t == null) {
                
            } else if (temp.visited) {
                res.add(temp.t.val);
            } else {
                temp.visited = true;
                s.push(temp);
                s.push(new Node(temp.t.right, false));
                s.push(new Node (temp.t.left, false));
            }
        }
        return res;
    }
    class Node{
        boolean visited;
        TreeNode t;
        public Node(TreeNode n, boolean flag) {
            t = n;
            visited = flag;
        }
    }
}

File: c++/145-BinaryTreePostorderTraversal.cpp
/*

Given a binary tree, return the postorder traversal of its nodes' values.

For example:
Given binary tree {1,#,2,3},
   1
    \
     2
    /
   3
return [3,2,1].

Note: Recursive solution is trivial, could you do it iteratively?

*/


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
// Recursive solutions 
class Solution1 {
public:
    void postorder(TreeNode* root, vector<int>& res){
        if (root == NULL) return ;
        postorder(root->left, res);
        postorder(root->right, res);
        res.push_back(root->val);
    }
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        postorder(root, res);
        return res;
    }
};

class Solution2 {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        if (root == NULL) return res;
        
        auto temp = postorderTraversal(root->left);
        res.insert(res.end(), temp.begin(), temp.end());
        temp = postorderTraversal(root->right);
        res.insert(res.end(), temp.begin(), temp.end());
        res.push_back(root->val);
        return res;
    }
};

// Iterative Solutions

// Morris Traversal
// Time:  O(n)
// Space: O(1)

class Solution3 {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        TreeNode dummy(INT_MIN);
        dummy.left = root;
        auto *cur = &dummy;
        while (cur) {
            if (!cur->left) {
                cur = cur->right;
            } else {
                auto *node = cur->left;
                while (node->right && node->right != cur) {
                    node = node->right;
                }
                if (!node->right) {
                    node->right = cur;
                    cur = cur->left;
                } else {
                    const auto& v = trace_back(cur->left, node);
                    res.insert(res.end(), v.cbegin(), v.cend());
                    node->right = nullptr;
                    cur = cur->right;
                }
            }
        }
        return res;
    }

private:
    vector<int> trace_back(const TreeNode *from, const TreeNode *to) {
        vector<int> res;
        auto *cur = from;
        while (cur != to) {
            res.emplace_back(cur->val);
            cur = cur->right;
        }
        res.emplace_back(to->val);
        reverse(res.begin(), res.end());
        return res;
    }
};

// Time:  O(n)
// Space: O(h)
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<pair<TreeNode *, bool>> s;
        s.emplace(root, false);
        while (!s.empty()) {
            bool visited;
            tie(root, visited) = s.top();
            s.pop();
            if (root == nullptr) {
                continue;
            }
            if (visited) {
                res.emplace_back(root->val);
            } else {
                s.emplace(root, true);
                s.emplace(root->right, false);
                s.emplace(root->left, false);
            }
        }
        return res;
    }
};
